java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,141 @@
+/**
+ * @name Temporary Directory Hijacking Vulnerability disclosure
+ * @description Detect temporary directory hijack vulnerability
+ * @kind path-problem
+ * @problem.severity error
+ * @precision very-high
+ * @id java/temp-directory-hijacking
+ */
+
+import java
+import semmle.code.java.controlflow.Guards
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.security.TempFileLib
+import DataFlow::PathGraph
+
+/**
+ * An expression that will create a directory without throwing an exception if a file/directory already exists.
+ */
+private predicate isNonThrowingDirectoryCreationExpression(Expr expr, MethodAccess creationCall) {
+  creationCall.getMethod() instanceof MethodFileCreatesDirs and creationCall.getQualifier() = expr
+}
+
+private class MethodFileDelete extends Method {
+  MethodFileDelete() {
+    getDeclaringType() instanceof TypeFile and
+    hasName("delete")
+  }
+}
+
+predicate isDeleteFileExpr(Expr expr) {
+  expr = any(MethodFileDelete m).getAReference().getQualifier()
+}
+
+private class TempDirHijackingToDeleteConfig extends TaintTracking::Configuration {
+  TempDirHijackingToDeleteConfig() { this = "TempDirHijackingToDeleteConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source.asExpr() =
+      any(MethodAccess ma |
+        ma.getMethod() instanceof MethodFileCreateTempFile and ma.getNumArgument() = 2
```

Comment:
Might be good to explain why check for `= 2`, that is, to match the method creating a file in the default OS temp directory.

Reply:
Done!